src/Tasks/SystemState.cs

```diff
@@ -548,12 +559,85 @@ out fileState.frameworkName
+        internal static SystemState DeserializePrecomputedCaches(string[] stateFiles, TaskLoggingHelper log, Type requiredReturnType, GetLastWriteTime getLastWriteTime, AssemblyTableInfo[] installedAssemblyTableInfo)
+        {
+            SystemState retVal = new SystemState();
+            retVal.SetGetLastWriteTime(getLastWriteTime);
+            retVal.SetInstalledAssemblyInformation(installedAssemblyTableInfo);
+            retVal.isDirty = stateFiles.Length > 0;
+            HashSet<string> assembliesFound = new HashSet<string>();
+
+            foreach (string stateFile in stateFiles)
+            {
+                // Verify that it's a real stateFile; log message but do not error if not
+                SystemState sfBase = (SystemState)DeserializeCache(stateFile, log, requiredReturnType, false);
+                foreach (string relativePath in sfBase.instanceLocalFileStateCache.Keys)
+                {
+                    if (!assembliesFound.Contains(relativePath))
+                    {
+                        FileState fileState = (FileState)sfBase.instanceLocalFileStateCache[relativePath];
+                        // Verify that the assembly is correct
+                        Guid mvid;
+                        string fullPath = Path.GetFullPath(Path.Combine(Path.GetDirectoryName(stateFile), relativePath));
+                        if (File.Exists(fullPath))
+                        {
+                            using (var reader = new PEReader(File.OpenRead(fullPath)))
+                            {
+                                var metadataReader = reader.GetMetadataReader();
+                                mvid = metadataReader.GetGuid(metadataReader.GetModuleDefinition().Mvid);
+                            }
+                            if (mvid.Equals(fileState.ModuleVersionID))
+                            {
+                                // Correct file path and timestamp
+                                fileState.LastModified = retVal.getLastWriteTime(fullPath);
+                                retVal.instanceLocalFileStateCache[fullPath] = fileState;
+                                assembliesFound.Add(relativePath);
+                            }
+                        }
+                    }
+                }
+            }
+
+            return retVal;
+        }
+
+        /// <summary>
+        /// Modifies this object to be more portable across machines, then writes it to stateFile.
+        /// </summary>
+        internal void SerializePrecomputedCache(string stateFile, TaskLoggingHelper log)
+        {
+            Hashtable newInstanceLocalFileStateCache = new Hashtable();
+            foreach (string path in instanceLocalFileStateCache.Keys)
+            {
+                // Add MVID to allow us to verify that we are using the same assembly later
+                FileState fileState = (FileState)instanceLocalFileStateCache[path];
```

Comment:
nit: Enumerate `instanceLocalFileStateCache` and not `instanceLocalFileStateCache.Keys` to avoid another hashtable lookup here.